### PR TITLE
Use dotnet format with CSharpier

### PR DESCRIPTION
**Pull request by pmachapman**
This Pull Request exists as a discussion point on using `dotnet format` with CSharpier.

### Overview

.NET 6.0 now includes a built in tool for code formatting called [dotnet format](https://github.com/dotnet/format).

This tool is more comprehensive than [CSharpier](https://csharpier.com/), and is completely configurable via .editorconfig, which means it applies any of your IDE code formatting warnings when the tool run.

`dotnet format` is lacking primarily in the regions of line length and spacing.

Although no tool comes close to prettier, and each tool has its pros and cons (e.g. CSharpier does line length, while `dotnet format` has many more options).

### Key Formatting Changes

 * Remove and sort usings
 * File level namespaces
 * Use expression bodies when possible
 * Use readonly fields when possible
 * Use expression bodies only when one line
 * Prefer using statements without curly braces

...and many more (see `.editorconfig` for details).

### Other Changes

 * Updating CSharpier to 0.22.1 (see release notes at https://github.com/belav/csharpier/releases/tag/0.22.1)
 * Updated the pre-commit hook to work with `dotnet format`

### FAQ

**Q. Why run `dotnet format style` and `dotnet format analyzers`?**
A. As per the documentation at https://csharpier.com/docs/IntegratingWithLinters, these two commands avoid the `dotnet format` rules that involve whitespace, and so there is no conflict between the output of `dotnet format` and CSharpier.

**Q. Wouldn't merging this make my currently open PR difficult to merge?**
A. Yes, it would. It is suggested that this PR is only merged if there are no other C# Pull Requests. Before merging, the commit ed4b20f0eaf8f29dc97bc6d00a8e923804540a99 should be dropped, `dotnet format` run from the repository root directory, and finally a new "C# modifications from dotnet format and CSharpier" commit created.

**Q. I don't like a specific formatting change - can we remove it?**
A. We can configure all of the changes made through the `.editorconfig` file. The best method would be to downgrade an option from `warning` to `suggestion` (to make it an IDE suggestion) or `silent` to suppress it. If we definitely want the inverse of the configuration option, then change the appropriate option to `true` or `false`.

**Q. How did you select what formatting rules to configure?**
I initially ran `dotnet format --severity info --report` on the solution to see what changes were made if all suggestions were implemented. I then enabled the matching rules to the suggestions, making sure I did not enable rules that `dotnet format` reported that it was unable to fix. The remaining rules I tweaked based on the [documentation](https://github.com/dotnet/format/blob/main/docs/Supported-.editorconfig-options.md).

**Q. Are any additional plugins required?**
A. No. Visual Studio uses `.editorconfig` for the IDE error list, and Visual Studio Code's C# extension is configured via `.vscode/settings.json`.

**Q. Why are there 100 CodeQL issues for this Pull Request?**
A. Because of the extensive reformatting of the source files. No new CodeQL issues should have been introduced that were not already present in the codebase. If this is an issue, we should resolve the CodeQL issues in the `master` branch before merging this Pull Request.

**Q. Help! This is impossible to review!**
A. To make it easier to review the scope and style of the C# code changes, I recommend selecting the "Hide whitespace" option when viewing the "Files Changed" tab in GitHub:
image

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1649)

### CODE REVIEW

**Review by marksvc — COMMENTED** [review]
I've looked through about half of the files. I have mixed thoughts about some of the different things that are being changed.

I think dotnet-format is more powerful, and I like what it can do for us. 

I don't like that it doesn't as aggressively pursue the line length limitation. Also, some basic formatting things seem omitted, like if I have code 
```cs
    public bool Thing(int a, int b, int elephantDogCamelHouseRabbit,

int shoeShirtHatHelmet, int redOrangeYellowGreenBlue, int oneTwoThreeFourFive)
    {
        return false;
    }
```
and I run `dotnet format`, it merely updates it to
```cs
    public bool Thing(int a, int b, int elephantDogCamelHouseRabbit,

int shoeShirtHatHelmet, int redOrangeYellowGreenBlue, int oneTwoThreeFourFive) => false;
```
If I run `dotnet csharpier` on it, it gets further cleaned up to
```cs
    public bool Thing(
        int a,
        int b,
        int elephantDogCamelHouseRabbit,
        int shoeShirtHatHelmet,
        int redOrangeYellowGreenBlue,
        int oneTwoThreeFourFive
    ) => false;
```
One of the great things about prettier is that it translates sloppily-formatted code into nicely-formatted code. CSharpier does a measure of this. I would be glad if we retained this sort of clean up if moving away from CSharpier.

I haven't kept up with the C# syntax changes that come out, and am not familiar with what is thought to be new and better ways to write the code. I wouldn't want us to not adopt a better syntax just because I am unfamiliar with it. 

I think some of the changes are ones I will just need to get used to (like including more content horizontally rather than vertically, and using `[2..3]` rather than `.Substring(2,3)`), and some that I will need to grow to appreciate (like writing small methods as lambdas (`private bool Thing() => false`) rather than in blocks (
```cs
{
  return false;
}
```

I have some questions, to help me understand some things:

Why would one make methods static, just because they _can_ be static? Is there some sort of performance gain or something?

Why would one make fields readonly, just because they _can_ be made readonly? Making fields readonly seems like it should be based on intent and a design, not based on whether it _can_ be made to be readonly. 

Why would one use lambdas (`private bool Thing() => false`) rather than write a method using blocks? As in:
```cs
private bool Thing()
{
    return false;
}
```
For this trivial example, `=> false` is nice and compact. But some instances are multiple (wrapped) lines, with different operations like type comparison, all as a single-line lambda. Are these lambdas thought to be overall easier to read (once I'm used to them)?


I think it will also be important to ask @ddaspit about his opinion on the changes.

File: src/SIL.XForge.Scripture/Services/ParatextSyncRunner.cs [hunk 1]
```diff
@@ -18,1589 +18,1553 @@
...
-        {
-            text = text.Trim().Replace("\r\n", "\n");
-            text = Regex.Replace(text, @"\n\s*<", "<", RegexOptions.CultureInvariant);
-            return XElement.Parse(text, LoadOptions.PreserveWhitespace);
-        }
+        var userIdsToRemove = new List<string>();
+        var projectUsers = await _realtimeService
+            .QuerySnapshots<User>()
+            .Where(u => _projectDoc.Data.UserRoles.Keys.Contains(u.Id) && u.ParatextId != null)
+            .Select(u => new { UserId = u.Id, u.ParatextId })
```
[the remark this answers is not in the document]
Comment (pmachapman) [reply src/SIL.XForge.Scripture/Services/ParatextSyncRunner.cs:1229]: It seems that C# is favoring name inference wherever possible. In some ways I see this as good - it encourages you to name your variables and properties consistently (and so your code is much clearer, and then can be less verbose). On the other...I am old school and like declaring everything explicitly :-)

**Comment by pmachapman** [comment]
Thank you for taking the time to look at this @marksvc - your comments are very helpful.

> I don't like that it doesn't as aggressively pursue the line length limitation. Also, some basic formatting things seem omitted

Me neither. I have been thinking on this, and wonder whether we should create a "csharp-format" script (perhaps both a .sh and .bat file) to run `dotnet format` then CSharpier? This might help us get the best of both worlds. It will mean that formatting is twice as slow, however.

> Why would one make methods static, just because they can be static? Is there some sort of performance gain or something?

Yes, there is a performance gain as the CLR does not have to check that the current object is not null. See https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1822 for details.

> Why would one make fields readonly, just because they can be made readonly? Making fields readonly seems like it should be based on intent and a design, not based on whether it can be made to be readonly.

`readonly` help with class immutability, but does not offer any performance enhancement. I find that making fields immutable helps me to not accidentally reassign them later. If I do need to reassign them, I have to consciously remove `readonly`, which makes me think about why I am doing that. But in short, yes, it is to do with intent and design.

> Why would one use lambdas (private bool Thing() => false) rather than write a method using blocks?

Really only for reasons of style. I have a background in functional programming, so like the lambda syntax. But it is something that is fully customizable in `.editorconfig`, and will not matter at the end of the day which way we go. When creating this PR, I did turn on *everything* I could think of to show some of the possibilities.

**Comment by marksvc** [comment]
> whether we should create a "csharp-format" script (perhaps both a .sh and .bat file) to run dotnet format then CSharpier? This might help us get the best of both worlds. It will mean that formatting is twice as slow, however.

Yes, that may be a good solution. As long as they don't clobber each other, and if things like IDE integration are still easy enough. We may find it's a lot easier to do IDE integration with an extension than with a script, for example.

> static ... performance gain

Okay I want to try going with that one then.

> field readonly 

Hmm. Well I am willing to try that. 

Problem using it:

I am finding that even though `dotnet format` does format a .cs file, I'm not having success getting it to be formatted from within VSCode. Such as from
- Ctrl+S
- Format Document
- Format Document (Forced)

I tried setting 
```
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "[csharp]": {
    "editor.defaultFormatter": "ms-dotnettools.csharp"
  },
```
or even
```
  "editor.defaultFormatter": "ms-dotnettools.csharp",
```
but without it formatting the file in the way we are trying out.

Also, when I make a local git commit, I see that it's doing something with lintstaged, but it does not end up actually formatting the file. 

<details>
<summary>For example, I see this lintstaged output.</summary>
<pre>
2022-12-21 12:54:06:  Staging: src/SIL.XForge/Services/ProjectService.cs
2022-12-21 12:54:06:  add 'src/SIL.XForge/Services/ProjectService.cs'
2022-12-21 12:54:22:  [update/dotnet-format b39cede8] test2
2022-12-21 12:54:22:   1 file changed, 1 insertion(+), 1 deletion(-)
2022-12-21 12:54:22:  [STARTED] Preparing lint-staged...
2022-12-21 12:54:22:  [SUCCESS] Preparing lint-staged...
2022-12-21 12:54:22:  [STARTED] Running tasks for staged files...
2022-12-21 12:54:22:  [STARTED] .lintstagedrc.yml — 1 file
2022-12-21 12:54:22:  [STARTED] *.{ts,js,json,css,scss,less,html,md,yml} — 0 files
2022-12-21 12:54:22:  [STARTED] *.cs — 1 file
2022-12-21 12:54:22:  [SKIPPED] *.{ts,js,json,css,scss,less,html,md,yml} — no files
2022-12-21 12:54:22:  [STARTED] dotnet format --include
2022-12-21 12:54:22:  [SUCCESS] dotnet format --include
2022-12-21 12:54:22:  [SUCCESS] *.cs — 1 file
2022-12-21 12:54:22:  [SUCCESS] .lintstagedrc.yml — 1 file
2022-12-21 12:54:22:  [SUCCESS] Running tasks for staged files...
2022-12-21 12:54:22:  [STARTED] Applying modifications from tasks...
2022-12-21 12:54:22:  [SUCCESS] Applying modifications from tasks...
2022-12-21 12:54:22:  [STARTED] Cleaning up temporary files...
2022-12-21 12:54:22:  [SUCCESS] Cleaning up temporary files...
</pre>
</details>

I'm using the "update/dotnet-format" branch. Is there something else I need to shake to get VSCode and my IDE to run the dotnet formatter?

**Comment by pmachapman** [comment]
> I'm using the "update/dotnet-format" branch. Is there something else I need to shake to get VSCode and my IDE to run the dotnet formatter?

I am noticing the same issue. It was less of a problem for me, as Visual Studio flagged the formatting issues as warnings for me to correct before I committed. It turns out that Omnisharp will use its built in formatter, not `dotnet format` (which will be called in the commit hook).

I can't find any way around it, but will update this PR to use both CSharpier and .NET format, and see how this works.

**Comment by pmachapman** [comment]
> Yes, that may be a good solution. As long as they don't clobber each other, and if things like IDE integration are still easy enough. We may find it's a lot easier to do IDE integration with an extension than with a script, for example.

In this updated PR, I have re-enabled CSharpier and its IDE integration, and have updated the `.lintstagedrc.yml` file to run CShapier after `dotnet format`.

I have been unable to find any way to execute `dotnet format` form the Format Document command in Visual Studio Code. This appears to be Omnisharp's own formatter, and does not allow a great deal of customization.

Finally, I noticed that `dotnet format` did not like the absolute paths being passed to it on Windows from lint-staged. To fix that, I have added an extra parameter `--relative` to the pre-commit hook, and it worked great for me for C# files and TypeScript files (i.e. it did not break prettier).

**Comment by marksvc** [comment]
Great; formatting on commit is working for me now. 
I had trouble getting csharpier to (go back to the previous behaviour of) triggering from saving a file but was able to get it going again by adding 
```json
  "[csharp]": {
    "editor.defaultFormatter": "csharpier.csharpier-vscode"
  },
```
to settings. 

I can't get VSCode to format with `dotnet format` either, nor success from searching. This seems like something that should be there, so maybe in the future it will. Tho that might not help the issue of wanting to use both dotnet format and csharpier, if that's still an issue at that time. Maybe there will be a combined solution by then.

**Comment by marksvc** [comment]
I like that the lambda method definitions are more vertically compact.
I don't like that the lambda method definitions don't let me anticipate that the content will always be on the next line, since the method content could be on the same line (and so my eyes don't know where to go! :). 
I don't think I would have gone with lambda method definitions if I were designing the style. But I am willing to try them, and they might grow on me. I'll be interested to hear what others say.

Regarding `foo.Substring(1,2)` being changed to `foo[1..2]`. It's a bit weird to me, but I am willing to try it.

Regarding applying `readonly` where possible, I don't like that this loses the difference between a field that is readonly because the class's implementation relies on it being so, and a field that is readonly just because nothing in the class so far modifies it. Deciding if `readonly` can be removed from a field in a small class may be easy to understand the implications. But when removing `readonly` from a field in a large class with lots of behaviour, it may be more challenging to know if there will be a bad result. Maybe not. The field probably wouldn't be being used all over the place.
It did appear that dotnet format only applied readonly to about 10 fields, from a quick look thru the PR.
I think I am okay with using "apply readonly if possible" if you think it will be helpful.

@ddaspit , @Nateowami , @RaymondLuong3 @nigel-wells , what do you guys think about the formatting changes?

**Review by RaymondLuong3 — COMMENTED** [review]
Regarding the `foo.Substring(1,2)` to `foo[1..2]` change, I do like that style. It reminds me of my days in python. I guess whether you like it depends on whether you have used this sort to substring syntax before.

Regarding the `readonly` where possible, I see both sides of the argument. I would lean toward not placing `readonly` just because you can since I like that when I see `readonly` in the code, I know that it was intentional, but I don't think I will be overly bothered by seeing `readonly` more often in the code.

> Reviewed 8 of 290 files at r1, 3 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1649#-:-NKsnSZx4kO1emZTWrfU:bd1dmp9)* status: 11 of 291 files reviewed, 1 unresolved discussion (waiting on @marksvc)

**Comment by pmachapman** [comment]
> Regarding applying readonly where possible, I don't like that this loses the difference between a field that is readonly because the class's implementation relies on it being so, and a field that is readonly just because nothing in the class so far modifies it. Deciding if readonly can be removed from a field in a small class may be easy to understand the implications. But when removing readonly from a field in a large class with lots of behaviour, it may be more challenging to know if there will be a bad result. 

>  I would lean toward not placing readonly just because you can since I like that when I see readonly in the code, I know that it was intentional, but I don't think I will be overly bothered by seeing readonly more often in the code.

I have downgraded `readonly` to just be an IDE suggestion, as I think you are both right - it is one of those changes that should be intentional by the developer.

**Review by ddaspit — COMMENTED** [review]
I don't really like the idea of running two different formatters on the code base. I do think it makes sense to have a linter and a formatter. The linter can be used to identify code smells and bad practices. It does not change the code. It only flags problems that the developer fixes. It is added as a check in the CI build. The formatter automatically fixes code to match the desired code style. I like to use opinionated formatters like CSharpier, Prettier, etc. On the Typescript side, we follow this convention: ESLint for linting and Prettier for formatting. .NET has code [quality analysis rules](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview?tabs=net-7#code-quality-analysis) built into the SDK, which could be used for linting.

Unfortunately, CSharpier doesn't support sorting and cleaning using statements, so it would be good to add that. The CSharpier [docs](https://csharpier.com/docs/IntegratingWithLinters) have more info on running dotnet-format and CSharpier together.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1649#-:-NL4cU7q-zyLnzsaLoPL:bsfimbs)* status: 10 of 291 files reviewed, 1 unresolved discussion (waiting on @marksvc and @RaymondLuong3)

**Review by pmachapman — COMMENTED** [review]
I agree - I don't like running two either. I can separate the linter for Visual Studio 2022 using the .editorconfig configuration - these appear as warnings, but can of course be customized to notices or errors. I cannot find a way to get these warnings or suggestions working fully in VSCode - a very small subset are supported by Omnisharp (C# for VSCode). I would think that they will eventually add support for this to Omnisharp, but I cannot find too much information on it beyond people requesting it.

Thank you for the documentation, Damien - I have updated the pre-commit hook to use `dotnet format` in the way that the CSharpier docs recommend, and  so the two programs no longer conflict with each other (the primary downside being that two commands have to be run for `dotnet format`).

Finally, I have also taken the opportunity to update CSharpier with this PR.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1649#-:-NM1KgLABrwv_zGNifxW:b4dxgep)* status: 7 of 291 files reviewed, 1 unresolved discussion (waiting on @marksvc and @RaymondLuong3)

**Comment by marksvc** [comment]
@pmachapman Sorry for not being very prompt in giving feedback on this. 

After considering what comments other have made, my thinking is that we should 
- Convert methods to static where we can
- Not require fields to be readonly
- For  `csharp_style_expression_bodied_methods` use `when_on_single_line`, so that we require the simplest methods to be expression bodies, but anything larger is in braces.

**Comment by marksvc** [comment]
Hmm, I don't think we have a single formatter option that still gives us these benefits. So it might just be good to benefit from using dotnet-format and csharpier together for now and we can see what the C# ecosystem provides in the future.

I find that as long as we don't include `dotnet format` or `dotnet format whitespace`, we at least don't conflict or flip-flop between dotnet-format and csharpier's formatting.


Interestingly, I find that running `dotnet format style && dotnet format analyzers` twice in a row can still result in some further changes. Likewise for a full `dotnet format`.
This.. might not happen very often in practice. But I can imagine that a workstation that runs `dotnet format style && dotnet format analyzers`, and pushes, may have the work flagged as not conforming to `dotnet format style && dotnet format analyzers` when the server checks if it has been run. So we can keep an eye out for that.

**Comment by pmachapman** [comment]
> For `csharp_style_expression_bodied_methods` use `when_on_single_line`, so that we require the simplest methods to be expression bodies, but anything larger is in braces.

I have made this update, and I really like it (it looks much better than when set to `true`). Thanks @marksvc!

> Interestingly, I find that running `dotnet format style && dotnet format analyzers` twice in a row can still result in some further changes.

I noticed this too when creating the commit with the changes in it. I too think it won't be too much of an issue in practice, but yes, something to look out for when making very large PRs.

**Comment by marksvc** [comment]
It's too bad we can't further set it to use expression bodies when the body would be 1-3 lines :-)

I'm good to go with it like this and we can revisit the settings if we want to make further changes. 

You're looking at the C# code quite a lot; is this set of settings something you'll be happy with? I'm glad to consider different settings.

I approve. You can decide on timing for when to merge and disrupt all the C# files :)

**Review by marksvc — APPROVED** [review]

**Review by marksvc — COMMENTED** [review]
> Reviewed 8 of 290 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1649)* status: 7 of 294 files reviewed, all discussions resolved (waiting on @RaymondLuong3)

**Comment by pmachapman** [comment]
> You're looking at the C# code quite a lot; is this set of settings something you'll be happy with? I'm glad to consider different settings.

I am totally happy with them. Having worked with a very wide range of codebases, I have found its doesn't matter what the formatting rules are so long as you you get adequate feedback during the development stage (via lint warnings, etc) that you have broken them.

> I approve. You can decide on timing for when to merge and disrupt all the C# files :)
 
Thank you! I will keep an eye on Pull Requests and Branches, and try to give a day's warning. Likely sometime next week-ish.

**Pull request merged**